### ISSUE TITLE
MM input causes unit problems

### ISSUE THREAD

**Issue by cschran** [cp2k/cp2k-input-tools#54]
I've tried to use cp2k-input-tools with a fairly standard MM Force-eval, but unfortunately there are problems with the `K` keywords in the `BEND` and `BOND` sections.

It seems that the parser tries to determine the standard unit for these variables, which are extracted from the `cp2k_input.xml` file. However, there are multiple definitions of `K` in there and it looks like the parser takes a definition with default unit `'internal_cp2k'`, which eventually triggers the following error:

```
in read_input_cp2k
    tree = parser.parse(f_in)
  File "/home/ucapcsc/software/miniconda3/lib/python3.7/site-packages/cp2k_input_tools/parser.py", line 216, in parse
    self._parse_as_keyword(line)
  File "/home/ucapcsc/software/miniconda3/lib/python3.7/site-packages/cp2k_input_tools/parser.py", line 149, in _parse_as_keyword
    kw = Keyword.from_string(kw_node, kw_value, self._key_trafo)  # the key_trafo is needed to mangle keywords
  File "/home/ucapcsc/software/miniconda3/lib/python3.7/site-packages/cp2k_input_tools/keyword_helpers.py", line 127, in from_string
    default_unit = UREG.parse_expression(default_unit)
  File "/home/ucapcsc/software/miniconda3/lib/python3.7/site-packages/pint/registry.py", line 1264, in parse_expression
    lambda x: self._eval_token(x, case_sensitive=case_sensitive, **values)
  File "/home/ucapcsc/software/miniconda3/lib/python3.7/site-packages/pint/pint_eval.py", line 102, in evaluate
    return define_op(self.left)
  File "/home/ucapcsc/software/miniconda3/lib/python3.7/site-packages/pint/registry.py", line 1264, in <lambda>
    lambda x: self._eval_token(x, case_sensitive=case_sensitive, **values)
  File "/home/ucapcsc/software/miniconda3/lib/python3.7/site-packages/pint/registry.py", line 1159, in _eval_token
    {self.get_name(token_text, case_sensitive=case_sensitive): 1}
  File "/home/ucapcsc/software/miniconda3/lib/python3.7/site-packages/pint/registry.py", line 643, in get_name
    raise UndefinedUnitError(name_or_alias)
pint.errors.UndefinedUnitError: 'internal_cp2k' is not defined in the unit registry
```

I have circumvented this problem by changing all `'internal_cp2k'` occasions in `cp2k_input.xml` to `'hartree'`, but this is a rather ugly fix...

**Comment by dev-zero** [comment]
Thanks for the bug report! Could you maybe attach a sample input? It does not have to be complete in the sense that one could run a calculation with it, just structurally valid with the problematic section in it.

The parser tries to normalize values here:

https://github.com/cp2k/cp2k-input-tools/blob/10aaa8697d916ee831897fa7a3429e596604d60e/cp2k_input_tools/keyword_helpers.py#L150-L154

The reason for this is that for serialization formats like json and yaml we yield values in the CP2K default unit since we do not export any unit specification.

It's a bit unfortunate that CP2K defines the internal units as abstract `internal_cp2k` rather than effective unit, especially for things where there would be a physical unit behind it (other than relative tolerances where this seems to be used otherwise).
This makes external linting of input files impossible since we could only validate such values by hardcoding the effective internal unit of those parameters within cp2k-input-tools.

I guess the proper way forward is to:

1. store an explicitly given unit for a quantity with a default unit of `internal_cp2k` and only validate it, do not attempt to convert the value from or to `internal_cp2k`.
2. when converting to a format (json or yaml) where the unit would be lost, either yield an error if the unit was explicitly given (since we can't convert it and the recipient of the json/yaml wouldn't know which unit it is), or yield the value as a string with the unit either suffixed (`1.23 angstrom`) or prefixed as in the CP2K input `[angstrom] 1.23`.

This would also delay unit conversion to the point when it becomes necessary (for linting it is not), rather than just attempting it all the time.

The intermediate solution: if the default unit is `internal_cp2k`, store as is, if there is an explicit unit given, raise an error since we can't convert to `internal_cp2k`.

**Comment by cschran** [comment]
Thanks for your fast response! This should trigger the bug:

```
import io
from cp2k_input_tools.parser import CP2KInputParserSimplified

parser = CP2KInputParserSimplified()

example_input = """
 &FORCE_EVAL
   METHOD  FIST
   &MM
     &FORCEFIELD
       IGNORE_MISSING_CRITICAL_PARAMS  T
       &BOND
         ATOMS O H
         KIND  HARMONIC
         K      4.7265512325474252E-01
         R0     1.9124028464802707E+00
       &END BOND
       &BEND
         ATOMS H O H
         KIND  HARMONIC
         K     1.2095435014802065E-01
         THETA0     1.9764108449583786E+00
       &END BEND
     &END FORCEFIELD
   &END MM
 &END FORCE_EVAL
"""

f = io.StringIO(example_input)
tree = parser.parse(f)
```

**Comment by dev-zero** [comment]
I just released v0.8.1 with a first fix (the basic solution I mentioned second in my previous comment) and opened #55 to track the enhancement which needs a bit more work.

**Comment by cschran** [comment]
great! thanks for this really fast solution. Cheers, Christoph
